karapace/utils.py

```diff
@@ -73,17 +84,34 @@ def default_json_serialization(
 SEPARATORS = (",", ":")
 
 
-def json_encode(obj, *, sort_keys: bool = True, binary=False, compact=False):
+def json_encode(obj, *, sort_keys: bool = True, binary=False, compact=False, indent: Optional[int] = None):
     separators = SEPARATORS if compact else None
+    arguments = {
+        "sort_keys": sort_keys,
+        "default": default_json_serialization,
+        "separators": separators,
+    }
+    if indent is not None:
+        arguments["indent"] = indent
     res = json.dumps(
         obj,
-        sort_keys=sort_keys,
-        default=default_json_serialization,
-        separators=separators,
+        **arguments,
     )
     return res.encode("utf-8") if binary else res
```

Comment:
Using overloads allows us to help editors in understanding what this function is going to return, as it either returns `bytes` or `str` in a predictable way. I also removed the global `SEPARATORS` constant, as there is no need to occupy the memory for it, as there is no reuse. Using `Optional[T] = None` for all arguments makes it possible to write wrapper functions, without copying the actual default values. JSON should also be compact by default, no need to waste precious bytes.

```python
@overload
def json_encode(
    it: Any,
    *,
    binary: Literal[False, None] = None,
    sort_keys: Optional[bool] = None,
    compact: Optional[bool] = None,
    indent: Optional[int] = None,
) -> str:
    ...


@overload
def json_encode(
    it: Any,
    *,
    binary: Literal[True],
    sort_keys: Optional[bool] = None,
    compact: Optional[bool] = None,
    indent: Optional[int] = None,
) -> bytes:
    ...


def json_encode(
    it: Any,
    *,
    binary: Optional[bool] = None,
    sort_keys: Optional[bool] = None,
    compact: Optional[bool] = None,
    indent: Optional[int] = None,
) -> Union[bytes, str]:
    kwargs = {}
    if indent is not None:
        kwargs["indent"] = indent
    if compact is not False and indent is None:  # compact by default
        kwargs["separators"] = (",", ":")
    if sort_keys is True:
        kwargs["sort_keys"] = True
    result = json.dumps(it, default=default_json_serialization, **kwargs)
    return result if binary is True else result.encode("utf8")
```

```python
string = json_encode({})  # correctly recognized as `str`
also_string = json_encode({}, binary=False)  # correctly recognized as `str`
binary = json_encode({}, binary=True)  # correctly recognized as `bytes`
```

Reply:
Done.